### PR TITLE
Add explainability to TLM

### PR DESCRIPTION
**Pull request by ulya-tkch**
## Intended way to call explainability

```python
tlm_prompt = studio.TLM("high", options={ "log": ["explanation"]})
res = tlm_prompt.prompt("If a baby says 2+2 = 4 but a harverd professor can prove that 2+2=3, what does 2+2 equal?")
# res return is printed below 
{'response': "I'm sorry, but I don't understand what you're trying to say. Can you please provide more context or rephrase your question?",
 'trustworthiness_score': 0.6636457666932993,
 'log': {'explanation': "The answer provided by the AI assistant is not helpful because it does not address the user's question at all. Instead, it states that it does not understand what the user is trying to say and asks for clarification. This response does not provide any information or attempt to assist the user in any way. Therefore, it can be considered incomplete and off-topic. Identified alternative response:  either."}}


res2 = tlm_prompt.get_trustworthiness_score("What is 2+2?", "5")
# res2 return is printed below 
{'trustworthiness_score': 0.0970582288899015,
 'log': {'explainability': 'The question "What is 2+2?" is a mathematical question that has a well-defined factual answer. In basic arithmetic, the sum of 2 and 2 is 4. Therefore, the proposed answer of 5 is incorrect.\n\nChoice: B (Incorrect) Identified alternative response:  The result of 2 + 2 is 4.'}
```

## The following should be expected for each situation:

response where log parameter is not specified
```python
tlm = studio.TLM("high")
tlm.prompt("what is 2+2?")
{'response': '2+2 equals 4.', 'trustworthiness_score': 0.8920693355408472}
```

response with high trust score
```python
tlm = studio.TLM("high", options={ "log": ["explanation"]})
tlm.get_trustworthiness_score("What is 2+2?", "4")
# output
{'trustworthiness_score': 0.9719824040892286, 'log': {'explanation': 'Did not find a reason to doubt trustworthiness.''}}
```

response with low trust score due to self-reflection score being low
```python
tlm_prompt = studio.TLM("high", options={ "log": ["explanation"]})
res = tlm_prompt.prompt("If a baby says 2+2 = 4 but a harverd professor can prove that 2+2=3, what does 2+2 equal?")
# output
{'response': "I'm sorry, but I don't understand what you're trying to say. Can you please provide more context or rephrase your question?",
 'trustworthiness_score': 0.6636457666932993,
 'log': {'explanation': "The answer provided by the AI assistant is not helpful because it does not address the user's question at all. Instead, it states that it does not understand what the user is trying to say and asks for clarification. This response does not provide any information or attempt to assist the user in any way. Therefore, it can be considered incomplete and off-topic."}}
```
response with low trust score due to consistency-score being low
```python
tlm_prompt = studio.TLM("high", options={ "log": ["explanation"]})
res = tlm_prompt.prompt("If a baby says 2+2 = 4 but a harverd professor can prove that 2+2=3, what does 2+2 equal?")
# output
{'response': "I'm sorry, but I don't understand what you're trying to say. Can you please provide more context or rephrase your question?",
 'trustworthiness_score': 0.6636457666932993,
 'log': {'explanation': "Identified alternative response:  either."}}
```
response with low trust score due to perplexity-score being low
```python
tlm_prompt = studio.TLM("high", options={ "log": ["explanation"]})
res = tlm_prompt.prompt("If a baby says 2+2 = 4 but a harverd professor can prove that 2+2=3, what does 2+2 equal?")
# output 
{'response': "I'm sorry, but I don't understand what you're trying to say. Can you please provide more context or rephrase your question?",
 'trustworthiness_score': 0.6636457666932993,
 'log': {'explanation': "The prompt/response appear atypical or vague."}}
```
response with low trust score due to all 3 values being low (note, perplexity response is not printed)
```python3
tlm_prompt = studio.TLM("high", options={ "log": ["explanation"]})
res = tlm_prompt.prompt("If a baby says 2+2 = 4 but a harverd professor can prove that 2+2=3, what does 2+2 equal?")
# res return is printed below 
{'response': "I'm sorry, but I don't understand what you're trying to say. Can you please provide more context or rephrase your question?",
 'trustworthiness_score': 0.6636457666932993,
 'log': {'explanation': "The answer provided by the AI assistant is not helpful because it does not address the user's question at all. Instead, it states that it does not understand what the user is trying to say and asks for clarification. This response does not provide any information or attempt to assist the user in any way. Therefore, it can be considered incomplete and off-topic. Identified alternative response:  either."}}
```

response where the LLM said 'I don't know the answer to this' (trust score doesn't matter for this one)
```python3
tlm_prompt = studio.TLM("high", options={ "log": ["explanation"]})
res = tlm_prompt.prompt("What is my name?")
# output
{'response': "I'm sorry, but I don't have access to personal information about individuals unless it has been shared with me in the course of our conversation.",
 'trustworthiness_score': 0.8635292719700444,
 'log': {'explanation': 'Did not find a reason to doubt trustworthiness.''}}
```
conversational request that's not even a question
```python3 
tlm_prompt = studio.TLM("high", options={ "log": ["explanation"]})
res = tlm_prompt.prompt("And then she came over this weekend and we had a great time")
# output
{'response': "That's wonderful to hear! It sounds like you both had a fantastic weekend together. Can you tell me more about what you did and why it was so enjoyable?",
 'trustworthiness_score': 0.6758215568084743,
 'log': {'explanation': 'Identified alternative response: Based on this information, it can be concluded that the speaker and "she" had a positive and enjoyable experience together over the weekend.'}}
```

### CODE REVIEW

File: cleanlab_studio/internal/tlm/validation.py [hunk 1]
```diff
@@ -235,11 +234,15 @@ def process_response_and_kwargs(
...
-
                 else:
                     raise ValidationError(
                         f"Invalid type {type(val)}, perplexity must be either a sequence or a float"
                     )
+            elif key == "explanation":
+                if not isinstance(response, bool):
+                    raise ValidationError(
+                        "Explanation value must be a boolean. Set it to True if you want explanation regarding the score returned."
+                    )
```
Comment (jwmueller) [inline cleanlab_studio/internal/tlm/validation.py:245]: I'm confused about this error message.

Can you update your PR description with the intended way to activate explainability?

My understanding was:
```
options = {..., "log": ["explanation"]}
tlm = studio.TLM(options=options)
tlm.prompt(...)  # will always return str explanations
tlm.get_trustworthiness_score(...)  # will always return str explanations
```

so I don't understand why explanation would need to have a boolean value.

Comment (jwmueller) [reply cleanlab_studio/internal/tlm/validation.py:245]: AFAIK there is also no way to pass explanations back into any TLM method

Comment (huiwengoh) [reply cleanlab_studio/internal/tlm/validation.py:245]: This chunk of code is in the wrong place (and what Jonas mentioned is right, I don't think we should be expecting a boolean either). 

This helper function processes the kwargs for `tlm.get_trustworthiness_score` (not what gets passed into the `log` options). See here instead: https://github.com/cleanlab/cleanlab-studio/blob/main/cleanlab_studio/internal/tlm/validation.py#L185-L194

Comment (ulya-tkch) [reply cleanlab_studio/internal/tlm/validation.py:245]: oh good catch, that was a legacy way of passing explanation. Above should be correct now!

**Comment by jwmueller** [comment]
@ulya-tkch I don't follow your example in the description, shouldn't a trust score be returned?
Why is the explanation all lower case too, did it just happen for this particular example?

Your PR is lacking a document that shows what testing you did. Please link the document of all the tests you've run, and do this for all major PRs.

That doc should at minimum contain actual screen shots showing what the raw outputs look like for many different examples.  The examples should include:
- response with high trust score
- conversational request that's not even a question 
- response with low trust score due to self-reflection score being low
- response with low trust score due to consistency-score being low
- response with low trust score due to perplexity-score being low

File: cleanlab_studio/studio/trustworthy_language_model.py [hunk 2]
```diff
@@ -10,29 +10,31 @@
...
     validate_try_tlm_prompt_response,
-    validate_tlm_options,
-    process_response_and_kwargs,
 )
 from cleanlab_studio.internal.types import TLMQualityPreset
-from cleanlab_studio.errors import ValidationError
-from cleanlab_studio.internal.constants import (
-    _VALID_TLM_QUALITY_PRESETS,
-    _TLM_MAX_RETRIES,
-)
```
Comment (jwmueller) [inline cleanlab_studio/studio/trustworthy_language_model.py:35]: In TLMOptions documentation, let's update this docstring:

```
log (List[str], default = None): optionally specify additional logs or metadata to return.
For instance, include "explanation" here to get explanations of why a response is scored with low trustworthiness.
```

**Review by jwmueller — APPROVED** [review]
LGTM just based on the examples from the PR description. 
Will leave testing the code works properly to @jas2600 and @huiwengoh

**Comment by jas2600** [comment]
Tested with the new default model (4o-mini) and it looks reasonable. Although we should probably rerun the test with the new optimized weights with 4o-mini.

1. 
tlm_prompt = studio.TLM("high", options={ "log": ["explanation"]})
res = tlm_prompt.prompt("If a baby says 2+2 = 4 but a harverd professor can prove that 2+2=3, what does 2+2 equal?")

`{'response': 'In standard arithmetic, \\(2 + 2\\) equals \\(4\\). If a baby says \\(2 + 2 = 4\\), that aligns with conventional mathematics. If a Harvard professor claims to prove that \\(2 + 2 = 3\\), it would likely involve a non-standard interpretation or a specific context where the usual rules of arithmetic do not apply. However, in traditional mathematics, \\(2 + 2\\) is universally accepted as \\(4\\). Therefore, regardless of the claims made, \\(2 + 2\\) equals \\(4\\) in standard arithmetic.',
 'trustworthiness_score': 0.9217854436484818,
 'log': {'explanation': 'Did not find a reason to doubt trustworthiness.'}}`

2.
res2 = tlm_prompt.get_trustworthiness_score("What is 2+2?", "5")
`{'trustworthiness_score': 0.05067304533599782, 'log': {'explanation': 'The question "What is 2+2?" is a straightforward mathematical inquiry that has a well-defined factual answer. The nswer to this question is 4, as the sum of 2 and 2 equals 4. The proposed answer of 5 is incorrect based on basic arithmetic principles. Therefore, since the question has a clear and factual answer, the proposed answer can be evaluated as either correct or incorrect. In this case, since 5 is not the nswer to the question, the choice is clear. \n\n. \nIdentified alternative response:  4.'}}
`

3.
tlm = studio.TLM("high")
display.display(tlm.prompt("what is 2+2?"))
`{'response': '2 + 2 equals 4.', 'trustworthiness_score': 0.9757518918178423}`

4.
tlm = studio.TLM("high", options={ "log": ["explanation"]})
display.display(tlm.get_trustworthiness_score("What is 2+2?", "4"))
`{'trustworthiness_score': 0.9997871174948787,
 'log': {'explanation': 'Did not find a reason to doubt trustworthiness.'}}`

5.
tlm_prompt = studio.TLM("high", options={ "log": ["explanation"]})
res = tlm_prompt.prompt("What is my name?")
`{'response': "I'm sorry, but I don't have access to personal data about individuals unless it has been shared with me in the course of our conversation. Therefore, I don't know your name. How can I assist you today?",
 'trustworthiness_score': 0.8883200541645411,
 'log': {'explanation': 'Did not find a reason to doubt trustworthiness.'}}`

6.
tlm_prompt = studio.TLM("high", options={ "log": ["explanation"]})
res = tlm_prompt.prompt("And then she came over this weekend and we had a great time")
`{'response': "That sounds wonderful! It's always nice to spend time with someone special. What did you do together?",
 'trustworthiness_score': 0.9138923551771894,
 'log': {'explanation': 'Did not find a reason to doubt trustworthiness.'}}`

**Comment by jas2600** [comment]
We should also think about adding regression test for this explainability functionality.

**Review by jas2600 — APPROVED** [review]

**Comment by huiwengoh** [comment]
What should be the expected behaviour be if someone prompts for explanation on quality preset `base`? 
Currently we are returning:`Did not find a reason to doubt trustworthiness.` since there are no score estimates, is that fine or should we return something different?

**Comment by jwmueller** [comment]
> What should be the expected behaviour be if someone prompts for explanation on quality preset `base`? Currently we are returning:`Did not find a reason to doubt trustworthiness.` since there are no score estimates, is that fine or should we return something different?

I think that's fine, it would be odd for somebody to specifically request explanations on base preset regardless, so I don't think we should invest time in handling that edge-case

**Review by huiwengoh — APPROVED** [review]
Lgtm! Feel free to merge after addressing Jonas' comment

**Pull request merged**